lib/bolt/project.rb

```diff
@@ -167,5 +168,132 @@ def check_deprecated_file
         Bolt::Logger.deprecation_warning('Using project.yaml instead of bolt-project.yaml', msg)
       end
     end
+
+    def migrate_project(configured_inventory_path, outputter)
+      migrating_inv = configured_inventory_path || inventory_file
+      inv_ok = inventory_1_to_2(migrating_inv, outputter) if migrating_inv.file?
+      config_ok = bolt_yaml_to_bolt_project(migrating_inv, outputter)
+      inv_ok && config_ok ? 0 : 1
+    end
+
+    # This could be made public and used elsewhere if the need arises
+    private def backup_file(origin_path, outputter)
+      unless File.exist?(origin_path)
+        outputter.print_message "Could not find file #{origin_path}, skipping backup."
+        return
+      end
+
+      date = Time.new.strftime("%Y%m%d_%H%M%S%L")
+      FileUtils.mkdir_p(backup_dir)
+
+      filename = File.basename(origin_path)
+      backup_path = File.join(backup_dir, "#{filename}.#{date}.bak")
+
+      outputter.print_message "Backing up #{filename} from #{origin_path} to #{backup_path}"
+
+      # No matter which error we rescue first, rubocop complains the order is wrong
+      # rubocop:disable Lint/ShadowedException
+      begin
+        FileUtils.cp(origin_path, backup_path)
+      rescue Errno::EACCES, StandardError => e
+        raise Bolt::FileError.new("#{e.message}; unable to create backup of #{filename}.", origin_path)
+      end
+      # rubocop:enable Lint/ShadowedException
+    end
+
+    private def bolt_yaml_to_bolt_project(migrating_inv, outputter)
+      # If bolt-project.yaml already exists
+      if project_file?
+        outputter.print_message "bolt-project.yaml already exists in Bolt "\
+          "project at #{path}. Skipping project file update."
+
+      # If bolt.yaml doesn't exist
+      elsif !config_file.file?
+        outputter.print_message "Could not find bolt.yaml in project at "\
+          "#{path}. Skipping project file update."
+
+      else
+        config_data = Bolt::Util.read_optional_yaml_hash(config_file, 'config')
+        transport_data, project_data = config_data.partition do |k, _|
+          Bolt::Config::INVENTORY_OPTIONS.keys.include?(k)
+        end.map(&:to_h)
+
+        if transport_data.any?
+          if File.exist?(inventory_file)
+            inventory_data = Bolt::Util.read_yaml_hash(migrating_inv, 'inventory')
+            updated_inventory = { 'config' => transport_data }.merge(inventory_data)
```

Comment:
We should deep merge `transport_data` with `inventory_data['config'] || {}` here, otherwise the config from the inventory will just override all of the config from `bolt.yaml`. For example, given these two files:

```yaml
# bolt.yaml
format: rainbow
ssh:
  user: bolt
  password: puppet
winrm:
  user: bolt
transport: docker
```
```yaml
# inventory.yaml
targets:
  - foo
  - bar
config:
  ssh:
    user: puppet
  transport: winrm
```

You will get the following inventory file:

```yaml
# inventory.yaml
---
config:
  ssh:
    user: puppet
  transport: winrm
targets:
- foo
- bar
```

Another benefit to merging the config sections separately is that the result can be assigned to `inventory_data['config']`, allowing us to preserve the order of the inventory file (at least those that weren't upgraded from v1).

Reply:
Oof, yeah, that would be bad :grimacing: